src/coreclr/debug/runtimeinfo/datadescriptor.h

```diff
@@ -0,0 +1,139 @@
+//   "cdacFieldtypeIdentifier" located at offset k in the type layout.  k is usually
+//   offsetof(SomeClass, m_FieldName) if the field is public
+//
+//     if the field is private, the convention is that SomeClass declares a friend struct
+//     cdac_offsets<T> and provides a specialization of cdac_offsets<T> with a public constexpr
+//     size_t member that holds the offset:
+//
+//     class MyClass {
+//       private:
+//         void* m_myField;
+//         friend template<typename T> cdac_offsets<T>;
+//     };
+//     template<> struct cdac_offsets<MyClass> {
+//       static constexpr size_t MyField = offsetof(MyClass, m_myField);
+//     };
+//
+//     then the field layout can be specified as
+//     CDAC_TYPE_FIELD(MyClassLayout, pointer, MyField, cdac_offsets<MyClass>::MyField)
+//  There can be zero or more CDAC_TYPE_FIELD entries per type layout
+//
+// CDAC_TYPE_END(cdacTypeIdentifier)  specifies the end of the type layout for cdacTypeIdentifier
+//
+// In <globals> the format is:
+//
+// CDAC_GLOBAL(cdacGlobalName, cdacTypeIdentifier, value)
+// or
+// CDAC_GLOBAL_POINTER(cdacGlobalName, cdacTypeIdentifier, address)
+//
+// Zero or more globals can be defined
+//
+// if a global is given with CDAC_GLOBAL(), `value` should be a constexpr uint64_t (or convertible
+// to uint64_t) for example, it can be a literal constant or a preprocessor definition
+//
+// if a global is a CDAC_GLOBAL_POINTER(), address should be a constexpr pointer or a constexpr
+// uintptr_t
+//
+//
+//
+// This file is compiled using the target architecture.  Preprocessor defines for the target
+// platform will be available.  It is ok to use `#ifdef`.
+
+#ifndef CDAC_BASELINE
+#define CDAC_BASELINE(identifier)
+#endif
+#ifndef CDAC_TYPES_BEGIN
+#define CDAC_TYPES_BEGIN()
+#endif
+#ifndef CDAC_TYPE_BEGIN
+#define CDAC_TYPE_BEGIN(tyname)
+#endif
+#ifndef CDAC_TYPE_SIZE
+#define CDAC_TYPE_SIZE(k)
+#endif
+#ifndef CDAC_TYPE_INDETERMINATE
+#define CDAC_TYPE_INDETERMINATE(tyname)
+#endif
+#ifndef CDAC_TYPE_FIELD
+#define CDAC_TYPE_FIELD(tyname,fieldtyname,fieldname,off)
+#endif
+#ifndef CDAC_TYPE_END
+#define CDAC_TYPE_END(tyname)
+#endif
+#ifndef CDAC_TYPES_END
+#define CDAC_TYPES_END()
+#endif
+#ifndef CDAC_GLOBALS_BEGIN
+#define CDAC_GLOBALS_BEGIN()
+#endif
+#ifndef CDAC_GLOBAL
+#define CDAC_GLOBAL(globalname,tyname,val)
+#endif
+#ifndef CDAC_GLOBAL_POINTER
+#define CDAC_GLOBAL_POINTER(globalname,addr)
+#endif
+#ifndef CDAC_GLOBALS_END
+#define CDAC_GLOBALS_END()
+#endif
+
+CDAC_BASELINE("empty")
+CDAC_TYPES_BEGIN()
+
+CDAC_TYPE_BEGIN(ManagedThread)
+CDAC_TYPE_INDETERMINATE(ManagedThread)
+CDAC_TYPE_FIELD(ManagedThread, GCHandle, GCHandle, cdac_offsets<Thread>::ExposedObject)
+CDAC_TYPE_FIELD(ManagedThread, pointer, LinkNext, cdac_offsets<Thread>::Link)
+CDAC_TYPE_END(ManagedThread)
+
+CDAC_TYPE_BEGIN(GCHandle)
+CDAC_TYPE_SIZE(sizeof(OBJECTHANDLE))
+CDAC_TYPE_END(GCHandle)
+
+CDAC_TYPES_END()
+
+CDAC_GLOBALS_BEGIN()
```

Comment:
I think their main use right now is just to group all the types and all the globals into separate collections.  I'm not sure if anything would go wrong if things were arranged differently, but since I didn't explicitly set things up to allow interleaving types and globals, I'm not sure what will happen. (the likeliest suspect is the names pool)